spring-integration-hazelcast/src/main/java/org/springframework/integration/hazelcast/leader/LeaderInitiator.java

```diff
@@ -0,0 +1,253 @@
+		if (this.running) {
+			this.running = false;
+			this.future.cancel(true);
+		}
+	}
+
+	/**
+	 * @return true if leadership election for this {@link #candidate} is running
+	 */
+	@Override
+	public boolean isRunning() {
+		return this.running;
+	}
+
+	@Override
+	public void afterPropertiesSet() throws Exception {
+		start();
+	}
+
+	@Override
+	public void destroy() throws Exception {
+		stop();
+		this.executorService.shutdown();
+	}
+
+	/**
+	 * Sets the {@link LeaderEventPublisher}.
+	 *
+	 * @param leaderEventPublisher the event publisher
+	 */
+	public void setLeaderEventPublisher(LeaderEventPublisher leaderEventPublisher) {
+		Assert.notNull(leaderEventPublisher);
+		this.leaderEventPublisher = leaderEventPublisher;
+		this.customPublisher = true;
+	}
+
+	/**
+	 * Callable that manages the acquisition of Hazelcast locks
+	 * for leadership election.
+	 */
+	class Initiator implements Callable<Void> {
+
+		@Override
+		public Void call() throws Exception {
+			Assert.state(LeaderInitiator.this.mapLocks != null);
+			HazelcastContext context = new HazelcastContext();
+			String role = LeaderInitiator.this.candidate.getRole();
+			boolean locked = false;
+
+			while (LeaderInitiator.this.running) {
+				try {
+					locked = LeaderInitiator.this.mapLocks.tryLock(role, Long.MAX_VALUE, TimeUnit.MILLISECONDS);
+					if (locked) {
+						LeaderInitiator.this.mapLocks.put(role, LeaderInitiator.this.candidate.getId());
+						LeaderInitiator.this.leaderEventPublisher.publishOnGranted(LeaderInitiator.this, context, LeaderInitiator.this.candidate.getRole());
+						LeaderInitiator.this.candidate.onGranted(context);
+						Thread.sleep(Long.MAX_VALUE);
+					}
+				}
+				catch (Exception e) {
+					// Catch and log any exceptions. This prevents the
+					// call method from exiting so that another attempt
+					// at acquiring leadership may be made.
+					// There is no need to reset the interrupt flag
+					// on InterruptedException because the interrupt
+					// is handled in the finally block below (just like
+					// any other exception)
+					logger.warn("Exception caught", e);
+				}
+				finally {
+					if (locked) {
+						LeaderInitiator.this.mapLocks.remove(role);
+						LeaderInitiator.this.mapLocks.unlock(role);
+						LeaderInitiator.this.candidate.onRevoked(context);
+						LeaderInitiator.this.leaderEventPublisher.publishOnRevoked(LeaderInitiator.this, context, LeaderInitiator.this.candidate.getRole());
+						locked = false;
+					}
+				}
+			}
+			return null;
+		}
+
+	}
+
+	/**
+	 * Implementation of leadership context backed by Hazelcast.
+	 */
+	class HazelcastContext implements Context {
+
+		@Override
+		public boolean isLeader() {
+			return LeaderInitiator.this.mapLocks != null && LeaderInitiator.this.mapLocks.isLocked(LeaderInitiator.this.candidate.getRole());
+		}
+
+		@Override
+		public void yield() {
+			if (LeaderInitiator.this.future != null) {
+				LeaderInitiator.this.future.cancel(true);
```

Comment:
Yeah... I see that the `yield` hasn't been fixed, as we discussion and had fixed in the `LockRegistryLeaderInitiator`: https://github.com/spring-projects/spring-integration/commit/1752204eb63c2edabde9c7fabb34cee76eb4535e
See commit comment since:
> Fix `yield()` logic based on the `Future.cancel(true)`.

I'll try to port my `yield()` here into Hazelcast module to prove the problem and than will try to fix it similar way via re-scheduling.